fri/src/layers.rs

```diff
@@ -0,0 +1,378 @@
+use crate::folder::MultiplicativeFriFolder;
+use anyhow::Error;
+use ark_ff::PrimeField;
+use ark_poly::EvaluationDomain;
+use std::sync::Arc;
+
+#[allow(dead_code)]
+pub trait FriLayer<F: PrimeField, E: EvaluationDomain<F>> {
+    fn get_layer_size(&self) -> usize;
+    fn get_domain(&self) -> E;
+    fn get_layer(&self) -> Result<Vec<F>, Error>;
+    fn eval_at_points(&self, required_indices: &[usize]) -> Vec<F>;
+    fn clone_arc(&self) -> Arc<dyn FriLayer<F, E>>;
+}
+
+#[derive(Clone)]
+pub struct FriLayerReal<F: PrimeField, E: EvaluationDomain<F>> {
+    layer_size: usize,
+    domain: E,
+    evaluation: Vec<F>,
+}
+
+#[allow(dead_code)]
+impl<F: PrimeField, E: EvaluationDomain<F>> FriLayerReal<F, E> {
+    pub fn new(domain: E, evaluation: Vec<F>) -> Self {
+        Self {
+            layer_size: domain.size(),
+            domain,
+            evaluation,
+        }
+    }
+
+    pub fn new_from_prev_layer(prev_layer: &dyn FriLayer<F, E>) -> Self {
+        Self {
+            layer_size: prev_layer.get_layer_size(),
+            domain: prev_layer.get_domain(),
+            evaluation: prev_layer.get_layer().unwrap(),
+        }
+    }
+}
+
+impl<F: PrimeField, E: EvaluationDomain<F> + 'static> FriLayer<F, E> for FriLayerReal<F, E> {
+    fn get_layer_size(&self) -> usize {
+        self.layer_size
+    }
+
+    fn get_domain(&self) -> E {
+        self.domain
+    }
+
+    fn get_layer(&self) -> Result<Vec<F>, Error> {
+        Ok(self.evaluation.clone())
+    }
+
+    fn eval_at_points(&self, required_indices: &[usize]) -> Vec<F> {
+        required_indices
+            .iter()
+            .map(|&i| self.evaluation[i])
+            .collect()
+    }
+
+    fn clone_arc(&self) -> Arc<dyn FriLayer<F, E>> {
```

Comment:
Why is this method needed? Can't the user just wrap the thing in an Arc himself?

This also copies the underlying struct, I would imagine that the goal was to have a number of references to the *same* immutable instance?

Reply:
Yes, removed this function.